cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,388 @@
+	if strategy.ProgressivePerGroup == nil {
+		strategy.ProgressivePerGroup = &RolloutProgressivePerGroup{}
+	}
+
+	// upgrade mandatory decision groups first
+	groupKeys := decisionGroupsToGroupKeys(strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
+
+	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutresult.ClustersToRollout) > 0 {
+		return &strategy, rolloutresult, nil
+	}
+
+	// parse timeout
+	failureTimeout, err := parseTimeout(strategy.ProgressivePerGroup.Timeout.Timeout)
+	if err != nil {
+		return &strategy, RolloutResult{}, err
+	}
+
+	// upgrade the rest decision groups
+	restClusterGroupKeys, restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(clusterGroupKeys)
+
+	rolloutresult = progressivePerGroup(restClusterGroupKeys, restClusterGroups, failureTimeout, statusFunc)
+	return &strategy, rolloutresult, nil
+}
+
+func progressivePerCluster(clusters []string, clusterToGroupKey map[string]clusterv1beta1.GroupKey, length int, timeout time.Duration, statusFunc ClusterRolloutStatusFunc) RolloutResult {
+	rolloutclusters := map[string]ClusterRolloutStatus{}
+	timeoutclusters := map[string]ClusterRolloutStatus{}
+
+	if length == 0 {
+		return RolloutResult{
+			ClustersToRollout: rolloutclusters,
+			ClustersTimeOut:   timeoutclusters,
+		}
+	}
+
+	// sort the clusters
+	sort.Strings(clusters)
+	for _, cluster := range clusters {
+		status := statusFunc(cluster)
+		if groupKey, exist := clusterToGroupKey[cluster]; exist {
+			status.GroupKey = groupKey
+		}
+
+		newstatus, needtorollout := needToRollout(status, timeout)
+		if needtorollout {
+			rolloutclusters[cluster] = status
+		}
+		if newstatus == TimeOut {
+			status.Status = newstatus
+			timeoutclusters[cluster] = status
+		}
+
+		if (len(rolloutclusters)%length == 0) && len(rolloutclusters) > 0 {
+			return RolloutResult{
+				ClustersToRollout: rolloutclusters,
+				ClustersTimeOut:   timeoutclusters,
+			}
+		}
+	}
+
+	return RolloutResult{
+		ClustersToRollout: rolloutclusters,
+		ClustersTimeOut:   timeoutclusters,
+	}
+}
+
+func progressivePerGroup(clusterGroupKeys []clusterv1beta1.GroupKey, clusterGroups map[clusterv1beta1.GroupKey]sets.Set[string], timeout time.Duration, statusFunc ClusterRolloutStatusFunc) RolloutResult {
+	rolloutclusters := map[string]ClusterRolloutStatus{}
+	timeoutclusters := map[string]ClusterRolloutStatus{}
+
+	for _, key := range clusterGroupKeys {
+		if subclusters, ok := clusterGroups[key]; ok {
+			// go through group by group
+			for _, cluster := range subclusters.UnsortedList() {
+				status := statusFunc(cluster)
+				status.GroupKey = key
+
+				newstatus, needtorollout := needToRollout(status, timeout)
+				if needtorollout {
+					rolloutclusters[cluster] = status
+				}
+				if newstatus == TimeOut {
```

Comment:
Why is `newstatus` only used if it's a timeout?

Reply:
When the input status is progressing or failed, and the status lasts longer than timeout, want to rewrite the status to timeout and return.

Reply:
I think `newstatus` always returns the correct status of either the same status or timeout. I think it may be better to just always use `newstatus`.